api/V1/recordingUtil.ts

```diff
@@ -143,6 +145,132 @@ export async function tryToMatchRecordingToStation(
+  if (!result) {
+    return;
+  }
+  let finished = false;
+  let currentFrame = 0;
+  let frame;
+  while (!finished) {
+    currentFrame++;
+    frame = await decoder.getNextFrame();
+    finished = frame == null;
+    if (currentFrame == frameNumber) {
+      break;
+    }
+  }
+  decoder.close();
+  return frame;
+}
+
+// Creates and saves a thumbnail for a recording using specified thumbnail info
+async function saveThumbnailInfo(recording, thumbnail) {
+  const frame = await getCPTVFrame(recording, thumbnail.frame_number);
+  const thumb = await createThumbnail(frame, thumbnail);
+
+  let upload = await modelsUtil
+    .openS3()
+    .upload({
+      Bucket: config.s3.bucket,
+      Key: `${recording.rawFileKey}-thumb`,
+      Body: thumb.data,
+      Metadata: thumb.meta
+    })
+    .promise()
+    .catch((err) => {
+      return err;
+    });
+  return upload;
+}
+
+// Create a png thumbnail image  from this frame with thumbnail info
+// Expand the thumbnail region such that it is a square and at least THUMBNAIL_MIN_SIZE
+// width and height
+//render the png in THUMBNAIL_PALETTE
+//returns {data: buffer, meta: metadata about image}
+async function createThumbnail(frame, thumbnail) {
+  const frameMeta = frame.meta.imageData;
+  const res_x = frameMeta.width;
+  const res_y = frameMeta.height;
+
+  const size = Math.max(THUMBNAIL_MIN_SIZE, thumbnail.height, thumbnail.width);
+  const thumbnail_data = new Uint8Array(size * size);
+
+  //dimensions to it is a square with a minimum size of THUMBNAIL_MIN_SIZE
+  const extra_width = (size - thumbnail.width) / 2;
+  thumbnail.x -= Math.ceil(extra_width);
+  thumbnail.x = Math.max(0, thumbnail.x);
+  thumbnail.width = size;
+  if (thumbnail.x + thumbnail.width > res_x) {
+    thumbnail.x = res_x - thumbnail.width;
+  }
+
+  const extra_height = (size - thumbnail.height) / 2;
+  thumbnail.y -= Math.ceil(extra_height);
+  thumbnail.y = Math.max(0, thumbnail.y);
+  thumbnail.height = size;
+  if (thumbnail.y + thumbnail.height > res_y) {
+    thumbnail.y = res_y - thumbnail.height;
+  }
+
+  let frame_start;
+  let thumb_index = 0;
+  for (let i = 0; i < size; i++) {
+    frame_start = (i + thumbnail.y) * res_x + thumbnail.x;
+    for (let offset = 0; offset < thumbnail.width; offset++) {
+      let pixel = frame.data[frame_start + offset];
+      pixel = (255 * (pixel - frameMeta.min)) / (frameMeta.max - frameMeta.min);
+      thumbnail_data[thumb_index] = pixel;
+      thumb_index++;
+    }
+  }
+  const frameBuffer = new Uint8ClampedArray(4 * thumbnail_data.length);
+  const { renderFrameIntoFrameBuffer, ColourMaps } = await dynamicImportESM(
+    "cptv-decoder"
+  );
+  let palette = ColourMaps[0];
+  for (const colourMap of ColourMaps) {
+    if (colourMap[0] == THUMBNAIL_PALETTE) {
+      palette = colourMap;
+    }
+  }
+  renderFrameIntoFrameBuffer(frameBuffer, thumbnail_data, palette[1], 0, 255);
+  const img = await sharp(frameBuffer, {
```

Comment:
We only resize if the thumbnail is smaller than the minimum size, but not if it is larger, right?

Reply:
yea i believe thats what "contain" fit does